File: data/recomposed_projects/llama-3-70b-instruct/body/joda-convert/src/main/org/joda/convert/AnnotationStringConverterFactory.py
from __future__ import annotations
import inspect
import re
import io
import typing
from typing import *
from src.main.org.joda.convert.ConstructorFromStringConverter import *
from src.main.org.joda.convert.FromString import *
from src.main.org.joda.convert.FromStringFactory import *
from src.main.org.joda.convert.MethodFromStringConverter import *
from src.main.org.joda.convert.ReflectionStringConverter import *
from src.main.org.joda.convert.StringConverter import *
from src.main.org.joda.convert.StringConverterFactory import *
from src.main.org.joda.convert.ToString import *
from src.main.org.joda.convert.TypedFromStringConverter import *


class AnnotationStringConverterFactory(StringConverterFactory):

    INSTANCE: AnnotationStringConverterFactory = None

    @staticmethod
    def initialize_fields() -> None:
        AnnotationStringConverterFactory.INSTANCE: AnnotationStringConverterFactory = (
            AnnotationStringConverterFactory()
        )

    def toString(self) -> str:
        return self.__class__.__name__

    def findConverter(
        self, cls: typing.Type[typing.Any]
    ) -> StringConverter[typing.Any]:
        return self.__findAnnotatedConverter(cls)  # capture generics

    def __eliminateEnumSubclass(
        self, cls: typing.Type[typing.Any]
    ) -> typing.Type[typing.Any]:
        sup = cls.__base__
        if sup is not None and sup.__base__ == Enum:
            return sup
        return cls

    def __findFromString(
        self, cls: typing.Type[typing.Any]
    ) -> typing.Union[inspect.Signature, typing.Callable]:
        methods = cls.__dict__.values()
        matched = None
        for method in methods:
            if not method.__name__ == "__init__" and not method.__name__ == "__new__":
                if hasattr(method, "__annotations__"):
                    if "FromString" in method.__annotations__:
                        if matched is not None:
                            raise RuntimeError(
                                "Two methods are annotated with @FromString: "
                                + cls.__name__
                            )
                        matched = method
                if hasattr(cls, "__annotations__"):
                    if "FromStringFactory" in cls.__annotations__:
                        if matched is not None:
                            raise RuntimeError(
                                "Class annotated with @FromString and @FromStringFactory: "
                                + cls.__name__
                            )
                        factory_methods = cls.__annotations__[
                            "FromStringFactory"
                        ].factory.__dict__.values()
                        for method in factory_methods:
                            if (
                                not method.__name__ == "__init__"
                                and not method.__name__ == "__new__"
                            ):
                                if cls in method.__annotations__:
                                    if hasattr(method, "__annotations__"):
                                        if "FromString" in method.__annotations__:
                                            if matched is not None:
                                                raise RuntimeError(
                                                    "Two methods are annotated with @FromString on the factory:"
                                                    + " "
                                                    + cls.__annotations__[
                                                        "FromStringFactory"
                                                    ].factory.__name__
                                                )
                                            matched = method
        return matched

    def __findFromStringMethod(
        self, cls: typing.Type[typing.Any], searchSuperclasses: bool
    ) -> TypedStringConverter[typing.Any]:
        loopCls = cls
        while loopCls is not None:
            fromString = self.__findFromString(loopCls)
            if fromString is not None:
                return MethodFromStringConverter(cls, fromString, loopCls)
            if searchSuperclasses == False:
                break
            loopCls = loopCls.__base__
        matched = None
        if searchSuperclasses:
            for loopIfc in self.__eliminateEnumSubclass(cls).__dict__.values():
                fromString = self.__findFromString(loopIfc)
                if fromString is not None:
                    if matched is not None:
                        raise RuntimeError(
                            "Two different interfaces are annotated with "
                            + "@FromString or @FromStringFactory: "
                            + cls.__name__
                        )
                    matched = MethodFromStringConverter(cls, fromString, loopIfc)
        return matched

    def __findFromStringConstructor(
        self, cls: typing.Type[typing.Any]
    ) -> TypedStringConverter[typing.Any]:

        pass  # LLM could not translate this method

    def __findToStringMethod(
        self, cls: typing.Type[typing.Any]
    ) -> typing.Union[inspect.Signature, typing.Callable]:
        matched = None
        loopCls = cls
        while loopCls is not None and matched is None:
            methods = loopCls.__dict__.values()
            for method in methods:
                if (
                    not method.__name__ == "__init__"
                    and not method.__name__ == "__new__"
                ):
                    toString = method.__annotations__.get("return")
                    if toString is not None:
                        if matched is not None:
                            raise RuntimeError(
                                "Two methods are annotated with @ToString: "
                                + cls.__name__
                            )
                        matched = method
            loopCls = loopCls.__base__
        if matched is None:
            for loopIfc in self.__eliminateEnumSubclass(cls).__bases__:
                methods = loopIfc.__dict__.values()
                for method in methods:
                    if (
                        not method.__name__ == "__init__"
                        and not method.__name__ == "__new__"
                    ):
                        toString = method.__annotations__.get("return")
                        if toString is not None:
                            if matched is not None:
                                raise RuntimeError(
                                    "Two methods are annotated with @ToString on interfaces: "
                                    + cls.__name__
                                )
                            matched = method
        return matched

    def __findAnnotatedFromStringConverter(
        self, cls: typing.Type[typing.Any]
    ) -> TypedStringConverter[typing.Any]:
        con = self.__findFromStringConstructor(cls)
        mth = self.__findFromStringMethod(
            cls, con is None
        )  # optionally checks superclasses
        if con is not None and mth is not None:
            raise RuntimeError(
                "Both method and constructor are annotated with @FromString: "
                + cls.__name__
            )
        return con if con is not None else mth

    def __findAnnotatedConverter(
        self, cls: typing.Type[typing.Any]
    ) -> StringConverter[typing.Any]:
        toString = self.__findToStringMethod(cls)  # checks superclasses
        if toString is None:
            return None
        fromString = self.__findAnnotatedFromStringConverter(cls)
        if fromString is None:
            raise RuntimeError(
                "Class annotated with @ToString but not with @FromString: "
                + cls.__name__
            )
        return ReflectionStringConverter(cls, toString, fromString)

    def __init__(self) -> None:
        pass

    def findFromStringConverter(
        self, cls: typing.Type[typing.Any]
    ) -> TypedStringConverter[typing.Any]:
        return self.__findAnnotatedFromStringConverter(cls)


AnnotationStringConverterFactory.initialize_fields()

File: java_projects/cleaned_reduced_projects/joda-convert/src/main/java/org/joda/convert/AnnotationStringConverterFactory.java
/*
 *  Copyright 2010-present Stephen Colebourne
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.joda.convert;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;

/**
 * Factory for {@code StringConverter} looking up annotations.
 *
 * <p>This class is immutable and thread-safe.
 *
 * @since 1.5
 */
final class AnnotationStringConverterFactory implements StringConverterFactory {

    /** Singleton instance. */
    static final AnnotationStringConverterFactory INSTANCE = new AnnotationStringConverterFactory();

    /** Restricted constructor. */
    private AnnotationStringConverterFactory() {}

    /**
     * Finds a converter by type.
     *
     * @param cls the type to lookup, not null
     * @return the converter, null if not found
     * @throws RuntimeException (or subclass) if source code is invalid
     */
    @Override
    public StringConverter<?> findConverter(Class<?> cls) {
        return findAnnotatedConverter(cls); // capture generics
    }

    /**
     * Finds a converter searching annotated.
     *
     * @param <T> the type of the converter
     * @param cls the class to find a method for, not null
     * @return the converter, not null
     * @throws RuntimeException if none found
     */
    private <T> StringConverter<T> findAnnotatedConverter(Class<T> cls) {
        Method toString = findToStringMethod(cls); // checks superclasses
        if (toString == null) {
            return null;
        }
        TypedFromStringConverter<T> fromString = findAnnotatedFromStringConverter(cls);
        if (fromString == null) {
            throw new IllegalStateException(
                    "Class annotated with @ToString but not with @FromString: " + cls.getName());
        }
        return new ReflectionStringConverter<T>(cls, toString, fromString);
    }

    /**
     * Finds a from-string converter by type.
     *
     * @param <T> the type of the converter
     * @param cls the type to lookup, not null
     * @return the converter, null if not found
     * @throws RuntimeException (or subclass) if source code is invalid
     */
    <T> TypedFromStringConverter<T> findFromStringConverter(Class<T> cls) {
        return findAnnotatedFromStringConverter(cls); // capture generics
    }

    /**
     * Finds a from-string converter.
     *
     * @param <T> the type of the converter
     * @param cls the class to find a method for, not null
     * @return the converter, null if not found
     * @throws RuntimeException if none found
     */
    private <T> TypedFromStringConverter<T> findAnnotatedFromStringConverter(Class<T> cls) {
        TypedFromStringConverter<T> con = findFromStringConstructor(cls);
        TypedFromStringConverter<T> mth =
                findFromStringMethod(cls, con == null); // optionally checks superclasses
        if (con != null && mth != null) {
            throw new IllegalStateException(
                    "Both method and constructor are annotated with @FromString: " + cls.getName());
        }
        return (con != null ? con : mth);
    }

    /**
     * Finds the conversion method.
     *
     * @param cls the class to find a method for, not null
     * @return the method to call, null means use {@code toString}
     * @throws RuntimeException if invalid
     */
    private Method findToStringMethod(Class<?> cls) {
        Method matched = null;
        Class<?> loopCls = cls;
        while (loopCls != null && matched == null) {
            Method[] methods = loopCls.getDeclaredMethods();
            for (Method method : methods) {
                if (!method.isBridge() && !method.isSynthetic()) {
                    ToString toString = method.getAnnotation(ToString.class);
                    if (toString != null) {
                        if (matched != null) {
                            throw new IllegalStateException(
                                    "Two methods are annotated with @ToString: " + cls.getName());
                        }
                        matched = method;
                    }
                }
            }
            loopCls = loopCls.getSuperclass();
        }
        if (matched == null) {
            for (Class<?> loopIfc : eliminateEnumSubclass(cls).getInterfaces()) {
                Method[] methods = loopIfc.getDeclaredMethods();
                for (Method method : methods) {
                    if (!method.isBridge() && !method.isSynthetic()) {
                        ToString toString = method.getAnnotation(ToString.class);
                        if (toString != null) {
                            if (matched != null) {
                                throw new IllegalStateException(
                                        "Two methods are annotated with @ToString on interfaces: "
                                                + cls.getName());
                            }
                            matched = method;
                        }
                    }
                }
            }
        }
        return matched;
    }

    /**
     * Finds the conversion method.
     *
     * @param <T> the type of the converter
     * @param cls the class to find a method for, not null
     * @return the method to call, null means none found
     * @throws RuntimeException if invalid
     */
    private <T> TypedFromStringConverter<T> findFromStringConstructor(Class<T> cls) {
        Constructor<T> con;
        try {
            con = cls.getDeclaredConstructor(String.class);
        } catch (NoSuchMethodException ex) {
            try {
                con = cls.getDeclaredConstructor(CharSequence.class);
            } catch (NoSuchMethodException ex2) {
                return null;
            }
        }
        FromString fromString = con.getAnnotation(FromString.class);
        if (fromString == null) {
            return null;
        }
        return new ConstructorFromStringConverter<T>(cls, con);
    }

    /**
     * Finds the conversion method.
     *
     * @param cls the class to find a method for, not null
     * @param searchSuperclasses whether to search superclasses
     * @return the method to call, null means not found
     * @throws RuntimeException if invalid
     */
    private <T> TypedFromStringConverter<T> findFromStringMethod(
            Class<T> cls, boolean searchSuperclasses) {
        Class<?> loopCls = cls;
        while (loopCls != null) {
            Method fromString = findFromString(loopCls);
            if (fromString != null) {
                return new MethodFromStringConverter<T>(cls, fromString, loopCls);
            }
            if (searchSuperclasses == false) {
                break;
            }
            loopCls = loopCls.getSuperclass();
        }
        TypedFromStringConverter<T> matched = null;
        if (searchSuperclasses) {
            for (Class<?> loopIfc : eliminateEnumSubclass(cls).getInterfaces()) {
                Method fromString = findFromString(loopIfc);
                if (fromString != null) {
                    if (matched != null) {
                        throw new IllegalStateException(
                                "Two different interfaces are annotated with "
                                        + "@FromString or @FromStringFactory: "
                                        + cls.getName());
                    }
                    matched = new MethodFromStringConverter<T>(cls, fromString, loopIfc);
                }
            }
        }
        return matched;
    }

    /**
     * Finds the conversion method.
     *
     * @param cls the class to find a method for, not null
     * @param matched the matched method, may be null
     * @return the method to call, null means not found
     * @throws RuntimeException if invalid
     */
    private Method findFromString(Class<?> cls) {
        Method[] methods = cls.getDeclaredMethods();
        Method matched = null;
        for (Method method : methods) {
            if (!method.isBridge() && !method.isSynthetic()) {
                FromString fromString = method.getAnnotation(FromString.class);
                if (fromString != null) {
                    if (matched != null) {
                        throw new IllegalStateException(
                                "Two methods are annotated with @FromString: " + cls.getName());
                    }
                    matched = method;
                }
            }
        }
        FromStringFactory factory = cls.getAnnotation(FromStringFactory.class);
        if (factory != null) {
            if (matched != null) {
                throw new IllegalStateException(
                        "Class annotated with @FromString and @FromStringFactory: "
                                + cls.getName());
            }
            Method[] factoryMethods = factory.factory().getDeclaredMethods();
            for (Method method : factoryMethods) {
                if (!method.isBridge() && !method.isSynthetic()) {
                    if (cls.isAssignableFrom(method.getReturnType())) {
                        FromString fromString = method.getAnnotation(FromString.class);
                        if (fromString != null) {
                            if (matched != null) {
                                throw new IllegalStateException(
                                        "Two methods are annotated with @FromString on the factory:"
                                                + " "
                                                + factory.factory().getName());
                            }
                            matched = method;
                        }
                    }
                }
            }
        }
        return matched;
    }

    private Class<?> eliminateEnumSubclass(Class<?> cls) {
        Class<?> sup = cls.getSuperclass();
        if (sup != null && sup.getSuperclass() == Enum.class) {
            return sup;
        }
        return cls;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName();
    }
}
